daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala

```diff
@@ -89,156 +95,184 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
-          case (TUnit, ValueUnit) =>
-            SValue.SUnit
-          case (TBool, ValueBool(b)) =>
-            if (b) SValue.SValue.True else SValue.SValue.False
-          case (TInt64, ValueInt64(i)) =>
-            SValue.SInt64(i)
-          case (TTimestamp, ValueTimestamp(t)) =>
-            SValue.STimestamp(t)
-          case (TDate, ValueDate(t)) =>
-            SValue.SDate(t)
-          case (TText, ValueText(t)) =>
-            SValue.SText(t)
-          case (TNumeric(TNat(s)), ValueNumeric(d)) =>
-            Numeric.fromBigDecimal(s, d).fold(fail, SValue.SNumeric)
-          case (TParty, ValueParty(p)) =>
-            SValue.SParty(p)
-          case (TContractId(_), ValueContractId(c)) =>
-            cids += c
-            SValue.SContractId(c)
-
-          // optional
-          case (TOptional(elemType), ValueOptional(mb)) =>
-            SValue.SOptional(mb.map((value: Value[ContractId]) => go(elemType, value, newNesting)))
-
-          // list
-          case (TList(elemType), ValueList(ls)) =>
-            SValue.SList(ls.map((value: Value[ContractId]) => go(elemType, value, newNesting)))
-
-          // textMap
-          case (TTextMap(elemType), ValueTextMap(entries)) =>
-            SValue.SGenMap(
-              isTextMap = true,
-              entries = entries.iterator.map { case (k, v) =>
-                SValue.SText(k) -> go(elemType, v, newNesting)
-              },
-            )
-
-          // genMap
-          case (TGenMap(keyType, valueType), ValueGenMap(entries)) =>
-            SValue.SGenMap(
-              isTextMap = false,
-              entries = entries.iterator.map { case (k, v) =>
-                go(keyType, k, newNesting) -> go(valueType, v, newNesting)
-              },
-            )
+        def newNesting = nesting + 1
+        def typeMismatch = fail(s"mismatching type: $ty and value: $value")
+        val (ty1, tyArgs) = destructApp(ty0)
+        ty1 match {
+          case TBuiltin(bt) =>
+            tyArgs match {
+              case Nil =>
+                (bt, value) match {
+                  case (BTUnit, ValueUnit) =>
+                    SValue.SUnit
+                  case (BTBool, ValueBool(b)) =>
+                    if (b) SValue.SValue.True else SValue.SValue.False
+                  case (BTInt64, ValueInt64(i)) =>
+                    SValue.SInt64(i)
+                  case (BTTimestamp, ValueTimestamp(t)) =>
+                    SValue.STimestamp(t)
+                  case (BTDate, ValueDate(t)) =>
+                    SValue.SDate(t)
+                  case (BTText, ValueText(t)) =>
+                    SValue.SText(t)
+                  case (BTParty, ValueParty(p)) =>
+                    SValue.SParty(p)
+                  case _ =>
+                    typeMismatch
+                }
+              case typeArg0 :: Nil =>
+                (bt, value) match {
+                  case (BTNumeric, ValueNumeric(d)) if typeArg0.isInstanceOf[TNat] =>
```

Comment:
I’m not that fond of `isInstanceOf`, can we pattern match instead?

Reply:
replaced.